packages/react-components/react-tree/src/utils/createHeadlessTree.ts

```diff
@@ -25,79 +27,146 @@ export type HeadlessTreeItem<Props extends HeadlessTreeItemProps> = {
+   * @param value the value of the item to get the ancestors from
+   */
+  ancestors(value: TreeItemValue): IterableIterator<HeadlessTreeItem<Props>>;
 };
 
 /**
  * creates a list of virtual tree items
  * and provides a map to access each item by id
  */
 export function createHeadlessTree<Props extends HeadlessTreeItemProps>(
-  virtualTreeItemProps: Props[],
+  initialProps: Props[] = [],
 ): HeadlessTree<Props> {
   const root = createHeadlessTreeRootItem();
   const itemsPerValue = new Map<TreeItemValue, HeadlessTreeItem<HeadlessTreeItemProps>>([[root.value, root]]);
-  const items: HeadlessTreeItem<HeadlessTreeItemProps>[] = [];
-
-  for (let index = 0; index < virtualTreeItemProps.length; index++) {
-    const { parentValue = virtualTreeRootId, ...treeItemProps } = virtualTreeItemProps[index];
-
-    const nextItemProps: Props | undefined = virtualTreeItemProps[index + 1];
-    const currentParent = itemsPerValue.get(parentValue);
-    if (!currentParent) {
-      if (process.env.NODE_ENV === 'development') {
-        // eslint-disable-next-line no-console
-        console.error(
-          `useHeadlessTree: item ${virtualTreeItemProps[index].value} is wrongly positioned, did you properly ordered provided item props? make sure provided items are organized`,
-        );
-      }
-      break;
-    }
-    const itemType =
-      treeItemProps.itemType ??
-      (treeItemProps.value === undefined || nextItemProps?.parentValue !== treeItemProps.value ? 'leaf' : 'branch');
-    const currentLevel = (currentParent.level ?? 0) + 1;
 
-    const virtualTreeItem: HeadlessTreeItem<HeadlessTreeItemProps> = {
-      value: treeItemProps.value,
-      getTreeItemProps: () => ({
-        ...treeItemProps,
-        'aria-level': currentLevel,
-        'aria-posinset': currentChildrenSize,
-        'aria-setsize': currentParent.childrenValues.length,
-        itemType,
-      }),
-      level: currentLevel,
-      parentValue,
-      childrenValues: [],
-      index: -1,
-    };
-    const currentChildrenSize = currentParent.childrenValues.push(virtualTreeItem.value);
-    itemsPerValue.set(virtualTreeItem.value, virtualTreeItem);
-    items.push(virtualTreeItem);
-  }
-
-  const virtualTreeItems: HeadlessTree<HeadlessTreeItemProps> = {
+  const headlessTree: HeadlessTree<HeadlessTreeItemProps> = {
     root,
-    size: items.length,
-    getByIndex: index => items[index],
+    get size() {
+      return itemsPerValue.size;
+    },
     getParent: key => itemsPerValue.get(itemsPerValue.get(key)?.parentValue ?? root.value) ?? root,
     get: key => itemsPerValue.get(key),
-    subtree: key => HeadlessTreeSubtreeGenerator(key, virtualTreeItems),
-    children: key => HeadlessTreeChildrenGenerator(key, virtualTreeItems),
-    ancestors: key => HeadlessTreeAncestorsGenerator(key, virtualTreeItems),
-    visibleItems: openItems => HeadlessTreeVisibleItemsGenerator(openItems, virtualTreeItems),
+    has: key => itemsPerValue.has(key),
+    add(props) {
```

Comment:
shouldn't the documentation be updated to use the new headless function?

Reply:
The `HeadlessTree` API is still internal for now, once this is better tested and stable I'll be sure to add it to `FlatHeadlessTree`  which is the external one.